tests/cpp/test_multidevice_matmul.cpp

```diff
@@ -436,4 +436,69 @@ TEST_F(DistributedMatmulTest, AnnotateWeightOnly) {
       __FILE__);
 }
 
+// linear([M, K], [N, K]) -> [M, N]
+//
+// K, the row dimension of the weight, is sharded on DIDx. Note that LinearOp's
+// weight is of shape [column, row]. This LinearOp is decomposed into a local
+// LinearOp followed by an Allreduce.
+TEST_F(DistributedMatmulTest, RowParallelLinear) {
+  auto fusion = std::make_unique<Fusion>();
+  FusionGuard fg(fusion.get());
+
+  TensorView* x = makeContigTensor(3);
+  TensorView* w = makeContigTensor(2);
+  TensorView* y = linear(x, w);
+  fusion->addInput(x);
+  fusion->addInput(w);
+  fusion->addOutput(y);
+
+  const auto d = communicator_->size();
+  auto mesh = DeviceMesh::createForNumDevices(d);
+
+  x->setDeviceMesh(mesh);
+  x->split(-1, d, /*inner_split=*/false);
+  x->axis(-2)->parallelize(ParallelType::DIDx);
+  x->setAllocationDomain(x->getLoopDomain(), true);
+
+  w->setDeviceMesh(mesh);
+  w->split(-1, d, /*inner_split=*/false);
+  w->axis(-2)->parallelize(ParallelType::DIDx);
+  w->setAllocationDomain(w->getLoopDomain(), true);
+
+  y->setDeviceMesh(mesh);
```

Comment:
This should not be needed. The final output is not sharded.

Reply:
> The final output is not sharded.

That's right, but...

`mesh` is required for replicated tensors as well as sharded tensors unless the whole program is single-GPU and doesn't contain **any** mesh. When `mesh` is not specified, sharding propagation kicks in and decides the mesh **and** the sharding. Because we haven't made sharding propagation to properly support DID loop split, I'm hesitant to skip the `mesh` here even if it works incidentally.